examples/arkode/C_serial/ark_damped_harmonic_symplectic.c

```diff
@@ -0,0 +1,322 @@
+/* clang-format off */
+/* ----------------------------------------------------------------------------
+ * Programmer(s): Cody J. Balos @ LLNL
+ * ----------------------------------------------------------------------------
+ * SUNDIALS Copyright Start
+ * Copyright (c) 2002-2023, Lawrence Livermore National Security
+ * and Southern Methodist University.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ * ----------------------------------------------------------------------------
+ * In this example we consider the time-dependent damped harmonic oscillator
+ *    q'(t) = p(t) exp(-F(t))
+ *    p'(t) = -(F(t) * p + omega^2(t) * q)
+ * With the initial conditions q(0) = 1, p(0) = 0.
+ * The Hamiltonian for the system is
+ *    H(p,q,t) = (p^2 * exp(-F(t)))/2 + (omega^2(t) * q^2 * exp(F(t)))/2
+ * where omega(t) = cos(t/2), F(t) = 0.018*sin(t/pi).
+ * We simulate the problem on t = [0, 30] using the symplectic methods in
+ * SPRKStep.
+ *
+ * This is example 7.2 from:
+ * Struckmeier, J., & Riedel, C. (2002). Canonical transformations and exact
+ * invariants for time‐dependent Hamiltonian systems. Annalen der Physik, 11(1),
+ * 15-38.
+ *
+ * The example has the following command line arguments:
+ *   --order <int>               the order of the method to use (default 4)
+ *   --dt <Real>                 the fixed-time step size to use (default 0.01)
+ *   --nout <int>                the number of output times (default 100)
+ *   --disable-tstop             turns off tstop mode
+ *   --use-compensated-sums      turns on compensated summation in ARKODE where
+ *                               applicable
+ * --------------------------------------------------------------------------*/
+/* clang-format on */
+
+#include <arkode/arkode_sprk.h>
+#include <arkode/arkode_sprkstep.h> /* prototypes for SPRKStep fcts., consts */
+#include <math.h>
+#include <nvector/nvector_serial.h> /* serial N_Vector type, fcts., macros  */
+#include <stdio.h>
+#include <string.h>
+#include <sundials/sundials_math.h> /* def. math fcns, 'sunrealtype'           */
+#include <sundials/sundials_nvector.h>
+#include <sundials/sundials_types.h>
+
+#include "arkode/arkode.h"
+
+#define PI SUN_RCONST(3.14159265358979323846264338327950)
+
+typedef struct
+{
+  int order;
+  int num_output_times;
+  int use_compsums;
+  int use_tstop;
+  sunrealtype Tf;
+  sunrealtype dt;
+} ProgramArgs;
+
+/* RHS functions */
+static int pdot(sunrealtype t, N_Vector y, N_Vector ydot, void* user_data);
+static int qdot(sunrealtype t, N_Vector y, N_Vector ydot, void* user_data);
+
+/* Helper functions */
+static sunrealtype Hamiltonian(N_Vector yvec, sunrealtype t);
+static int ParseArgs(int argc, char* argv[], ProgramArgs* args);
+static void PrintHelp();
+static int check_retval(void* returnvalue, const char* funcname, int opt);
```

Comment:
Here and in the other examples I think it would be good to move utilities like `ParseArgs`, `PrintHelp`, `check_retval`, etc. to a header file for the example to streamline what's in the source to just what a user needs to provide.

Reply:
Are you suggesting an example specific header file like `ark_damped_harmonic_symplectic.h`?

Reply:
Yes, that's what I was thinking (in particular for things like parsing input args that can be lengthy but not that important to understanding how to use the code)